ops/hookcmds.py

```diff
@@ -0,0 +1,995 @@
+            protocol = None
+        if '-' in port:
+            port, to_port = port.split('-')
+            to_port = int(to_port)
+        else:
+            to_port = None
+        if port == 'icmp':
+            protocol = port
+            port = None
+        else:
+            port = int(port)
+        assert protocol in ('tcp', 'udp', 'icmp')
+        ports.append(Port(protocol=protocol, port=port, to_port=to_port))
+    return ports
+
+
+@overload
+def relation_get(
+    id: int | None = None,
+    *,
+    unit: str | None = None,
+    app: bool = False,
+) -> dict[str, str]: ...
+@overload
+def relation_get(
+    id: int | None = None,
+    *,
+    key: str,
+    unit: str | None = None,
+    app: bool = False,
+) -> str: ...
+def relation_get(
+    id: int | None = None,
+    *,
+    key: str | None = None,
+    unit: str | None = None,
+    app: bool = False,
+) -> dict[str, str] | str:
+    """Get relation settings.
+
+    Note that ``id`` can only be ``None`` if the current hook is a relation
+    event, in which case Juju will use the ID of the relation that triggered the
+    event.
+    """
+    args = ['relation-get']
+    if id is not None:
+        args.extend(['-r', str(id)])
+    if app:
+        args.append('--app')
+    if key is not None:
+        args.append(key)
+    if unit is not None:
+        args.append(unit)
+    result = _run(*args, return_output=True, use_json=True)
+    if key is not None:
+        return cast('str', result)
+    return cast('dict[str, str]', result)
+
+
+def relation_ids(name: str) -> list[str]:
+    """List all relation IDs for the given endpoint."""
+    result = _run('relation-ids', name, return_output=True, use_json=True)
+    return cast('list[str]', result)
+
+
+def relation_list(id: int | None = None, *, app: bool = False) -> list[str]:
+    """List relation units.
+
+    Note that ``id`` can only be ``None`` if the current hook is a relation
+    event, in which case Juju will use the ID of the relation that triggered the
+    event.
+    """
+    args = ['relation-list']
+    if app:
+        args.append('--app')
+    if id is not None:
+        args.extend(['-r', str(id)])
+    result = _run(*args, return_output=True, use_json=True)
+    return cast('list[str]', result)
+
+
+def relation_model_get(id: int | None = None) -> RelationModel:
+    """Get details about the model hosting a related application.
+
+    Note that ``id`` can only be ``None`` if the current hook is a relation
+    event, in which case Juju will use the ID of the relation that triggered the
+    event.
+    """
+    args = ['relation-model-get']
+    if id is not None:
+        args.extend(['-r', str(id)])
+    result = cast('_RelationModelDict', _run(*args, return_output=True, use_json=True))
+    return RelationModel(uuid=uuid.UUID(result['uuid']))
+
+
+def relation_set(
+    data: Mapping[str, str],
+    id: int | None = None,
+    *,
+    app: bool = False,
+    file: pathlib.Path | None = None,
```

Comment:
I don't think we should support *file*. It's not "real" anyway (we're loading the file on the Python side), and it just updates the data dict, only to send all the values via stdin. So I think we just drop this parameter.